#166840937 User can Bookmark/UnBookmark articles

tolumide-ng opened the pull request:
#### Description
Users can bookmark an article  via **/api/v1/articles/bookmark** and unbookmark **/api/v1/articles/unbookmark** respectively so far they both are registered users of the errorSwag application 

#### Type of Change
- [x] New feature (Non-breaking change which adds functionality)
- [x] Bug fix (Non-breaking change which adds functionality)


#### How Has This Been Tested?
- [x] Integration Testing


#### Pivotal Stories
[#166840937](https://www.pivotaltracker.com/story/show/166840937)

Onyimatics commented on src/controllers/article.controller.js:
```diff
@@ -0,0 +1,88 @@
+   * @static
+   */
+  static async getBookMarks(req, res) {
+    // const { articleId } = req.body;
```
please kindly remove this if you are not using it. @tolumide-ng

tolumide-ng replied:
Thank you, I would remove this

Onyimatics commented on src/test/_user.spec.js:
```diff
@@ -132,9 +132,9 @@ describe('Test user login, signup and account verification', () => {
     });
 
-    it('it should login the user into an account', done => {
+    it.skip('it should login the user into an account', done => {
```
I guess you forgot to remove this `skip`.

tolumide-ng replied:
No, I did not forget to remove this, the test is actually dependent which isn't supposed to be. Hence it fails when it should not. @timi-codes as promised to work on this.

kechyy replied:
> No, I did not forget to remove this, the test is actually dependent which isn't supposed to be. Hence it fails when it should not. @timi-codes as promised to work on this.

@tolumide-ng Don't you think it will be nice you discuss with @timi-codes on this rather than skipping it?

tolumide-ng replied:
Thank you @kechyy , I have discussed with @timi-codes about this and he has promised to make a fix for the test that wouldn't be dependent in his next PR

Onyimatics reviewed (changes requested):
Looks good to me! I like the way you wrote the tests. Please kindly make the changes I pointed out.

tolumide-ng commented:
> Looks good to me! I like the way you wrote the tests. Please kindly make the changes I pointed out.

Thank you for taking your time to review this. I would implement the changes as soon as possible.😊

segunolalive commented on src/middleware/article.middleware.js:
```diff
@@ -0,0 +1,21 @@
+    return responseGenerator.sendError(
+      res,
+      400,
+      `There is no article with this Id = ${articleId}`
```
The requested article was not found

tolumide-ng replied:
Change implemented. Thank you

segunolalive commented on src/test/article.spec.js:
```diff
@@ -0,0 +1,132 @@
+    const firstUser = await createUser();
+    const secondUser = await createUser();
+    const theArticle = await createArticle(
+      await getArticle({ authorId: secondUser.id })
```
`generateArticle` might be better here

segunolalive commented on src/test/article.spec.js:
```diff
@@ -0,0 +1,132 @@
+      await getArticle({ authorId: secondUser.id })
+    );
+    const numberOfArticles = await BaseRepository.findAll(db.Article);
+    const numberOfBookmarks = await BaseRepository.findAll(db.Bookmark, {
```
You can use sequelize `count`. You can rename to `bookmarksCount`

tolumide-ng replied:
Change implemented thank you

segunolalive commented on src/test/article.spec.js:
```diff
@@ -0,0 +1,132 @@
+    const newNumberOfBookmarks = await BaseRepository.findAll(db.Bookmark, {
+      userId: firstUser.id
+    });
+    expect(newNumberOfBookmarks.length).to.equal(1);
```
Assert that the article id sent matches what was inserted into the db

tolumide-ng replied:
Change implemented, thank you

segunolalive commented on src/test/article.spec.js:
```diff
@@ -0,0 +1,132 @@
+    const res = await server()
+      .get(`${ARTICLES_API}/bookmark`)
+      .set('token', token);
+    expect(res.status).to.equal(200);
```
Assert that the data returned matches what's expected.

segunolalive reviewed (changes requested):
Good job. Make the changes.

Onyimatics commented on src/database/models/article.js:
```diff
@@ -0,0 +1,31 @@
+  Article.associate = models => {
+    Article.belongsTo(models.User, {
+      through: 'Articles',
+      // as: 'users',
```
please remove this line if you are not using it. @tolumide-ng

Onyimatics reviewed (changes requested):
Good job @tolumide-ng. Please kindly attend to the comments I left. Thank you

tolumide-ng committed:
feat(users): Users can Bookmark articles

- User can bookmark an article to read later
- User can unbookmark articles
- User can fetch all bookmarked articles
- Creates findAndInclude method in baseRepository
- Removes duplicate update method on baserepository

[Delivers #166840937]

Onyimatics reviewed (approved):
Good job @tolumide-ng 
I like the way you named the variables you used.

kechyy reviewed (changes requested):
Good Job @tolumide-ng . Kindly attend to the one comment I left for you. 
Thank you.

kechyy reviewed (approved):
Good Job @tolumide-ng . Kindly attend to the one comment I left for you. 
Thank you.

tolumide-ng commented:
> Good Job @tolumide-ng . Kindly attend to the one comment I left for you.
> Thank you.

Thank you 👍
